va/win32/va_win32.c

```diff
@@ -0,0 +1,329 @@
+    VADriverContextP  pDriverContext;
+
+    pDisplayContext = va_newDisplayContext();
+    if (!pDisplayContext)
+        return NULL;
+
+    pDisplayContext->vaIsValid       = va_DisplayContextIsValid;
+    pDisplayContext->vaDestroy       = va_DisplayContextDestroy;
+    pDisplayContext->vaGetDriverNameByIndex = va_DisplayContextGetDriverNameByIndex;
+    pDisplayContext->vaGetNumCandidates = va_DisplayContextGetNumCandidates;
+    pDisplayContext->opaque = calloc(1, sizeof(VADisplayContextWin32));
+
+    VADisplayContextWin32* pWin32Ctx = (VADisplayContextWin32*) pDisplayContext->opaque;
+    sprintf(pWin32Ctx->default_driver_name, "%s", "vaon12");
+    fprintf(stderr, "VA_Win32: Setting default driver name %s\n", pWin32Ctx->default_driver_name);
+
+    if (adapter_luid) {
+        // Copy LUID information to display context
+        memcpy(&pWin32Ctx->adapter_luid, adapter_luid, sizeof(pWin32Ctx->adapter_luid));
+
+        // Load the preferred driver name from the driver registry if available
+        fprintf(stderr, "VA_Win32: Attempting driver registry discovery for LUID: %ld %ld \n", pWin32Ctx->adapter_luid.LowPart, pWin32Ctx->adapter_luid.HighPart);
+        pWin32Ctx->registry_driver_available_flag = TryLoadDriverNameFromRegistry(&pWin32Ctx->adapter_luid, pWin32Ctx) ? 1 : 0;
+        if (pWin32Ctx->registry_driver_available_flag) {
+            fprintf(stderr, "VA_Win32: Found driver %s in the registry for LUID %ld %ld \n", pWin32Ctx->registry_driver_name, pWin32Ctx->adapter_luid.LowPart, pWin32Ctx->adapter_luid.HighPart);
+        }
+        else
+        {
+            fprintf(stderr, "VA_Win32: Couldn't find a driver in the registry for LUID %ld %ld \n", pWin32Ctx->adapter_luid.LowPart, pWin32Ctx->adapter_luid.HighPart);
+        }
+    }
+
+    pDriverContext = va_newDriverContext(pDisplayContext);
+    if (!pDriverContext) {
+        va_DisplayContextDestroy(pDisplayContext);
+        return NULL;
+    }
+
+    pDriverContext->display_type = VA_DISPLAY_WIN32;
+
+    if (adapter_luid) {
+        // Copy LUID information to driver context
+        pDriverContext->native_dpy   = calloc(1, sizeof(*adapter_luid));
+        memcpy(pDriverContext->native_dpy, adapter_luid, sizeof(*adapter_luid));
+    }
+
+    return (VADisplay)pDisplayContext;
+}
+
+VAStatus vaPutSurface(
+    VADriverContextP ctx,
+    VASurfaceID surface,
+    HWND draw,
+    short srcx,
+    short srcy,
+    unsigned short srcw,
+    unsigned short srch,
+    short destx,
+    short desty,
+    unsigned short destw,
+    unsigned short desth,
+    VARectangle *cliprects, /* client supplied clip list */
+    unsigned int number_cliprects, /* number of clip rects in the clip list */
+    unsigned int flags /* de-interlacing flags */
+)
+{
+    return ctx->vtable->vaPutSurface(ctx, surface, (void *)draw, srcx, srcy, srcw, srch,
+                                     destx, desty, destw, desth,
+                                     cliprects, number_cliprects, flags);
```

Comment:
I would encourage you to implement vaExportSurfaceHandle and drop this PutSurface together.

The former should (implementation permitting) provide zero copy path, while does is always a copy. In addition, the API is defined only on va-x11 + the only project that I could find using it (libva-utils) simply does not work.

Aka it doesn't seem like a good choice from efficiency, portability and acceptance (aka actual users) POV.

Reply:
Thanks for pointing this out. Sounds better/simpler to me too that apps use vaExportSurfaceHandle (ideally with no copies) and implement their own/defer to other present codepaths. If for some reason in the future a customer wants vaPutSurface for a good reason, we can extend the win32 API at that point.